pkg/controller/persistentvolumelabel/pvlcontroller.go

```diff
@@ -0,0 +1,368 @@
+		}
+		if volume.Spec.GCEPersistentDisk != nil {
+			labels, err := pvlc.findGCEPDLabels(volume)
+			if err != nil {
+				return fmt.Errorf("error querying GCE PD volume %s: %v", volume.Spec.GCEPersistentDisk.PDName, err)
+			}
+			volumeLabels = labels
+		}
+		return pvlc.updateVolume(volume.Name, volumeLabels)
+	}
+
+	return nil
+}
+
+func (pvlc *PersistentVolumeLabelController) findAWSEBSLabels(volume *v1.PersistentVolume) (map[string]string, error) {
+	// Ignore any volumes that are being provisioned
+	if volume.Spec.AWSElasticBlockStore.VolumeID == vol.ProvisionedVolumeName {
+		return nil, nil
+	}
+	ebsVolumes, err := pvlc.getEBSVolumes()
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: GetVolumeLabels is actually a method on the Volumes interface
+	// If that gets standardized we can refactor to reduce code duplication
+	spec := aws.KubernetesVolumeID(volume.Spec.AWSElasticBlockStore.VolumeID)
+	labels, err := ebsVolumes.GetVolumeLabels(spec)
+	if err != nil {
+		return nil, err
+	}
+
+	return labels, nil
+}
+
+// getEBSVolumes returns the AWS Volumes interface for ebs
+func (pvlc *PersistentVolumeLabelController) getEBSVolumes() (aws.Volumes, error) {
+	pvlc.mutex.Lock()
+	defer pvlc.mutex.Unlock()
+
+	if pvlc.ebsVolumes == nil {
+		awsCloudProvider, ok := pvlc.cloud.(*aws.Cloud)
+		if !ok {
+			// GetCloudProvider has gone very wrong
+			return nil, fmt.Errorf("error retrieving AWS cloud provider")
+		}
+		pvlc.ebsVolumes = awsCloudProvider
+	}
+	return pvlc.ebsVolumes, nil
+}
+
+func (pvlc *PersistentVolumeLabelController) findGCEPDLabels(volume *v1.PersistentVolume) (map[string]string, error) {
+	// Ignore any volumes that are being provisioned
+	if volume.Spec.GCEPersistentDisk.PDName == vol.ProvisionedVolumeName {
+		return nil, nil
+	}
+
+	provider, err := pvlc.getGCECloudProvider()
+	if err != nil {
+		return nil, err
+	}
+
+	// If the zone is already labeled, honor the hint
+	zone := volume.Labels[kubeletapis.LabelZoneFailureDomain]
+
+	labels, err := provider.GetAutoLabelsForPD(volume.Spec.GCEPersistentDisk.PDName, zone)
+	if err != nil {
+		return nil, err
+	}
+
+	return labels, nil
+}
+
+// getGCECloudProvider returns the GCE cloud provider, for use for querying volume labels
+func (pvlc *PersistentVolumeLabelController) getGCECloudProvider() (*gce.GCECloud, error) {
+	pvlc.mutex.Lock()
+	defer pvlc.mutex.Unlock()
+
+	if pvlc.gceCloudProvider == nil {
+		gceCloudProvider, ok := pvlc.cloud.(*gce.GCECloud)
+		if !ok {
+			// GetCloudProvider has gone very wrong
+			return nil, fmt.Errorf("error retrieving GCE cloud provider")
+		}
+		pvlc.gceCloudProvider = gceCloudProvider
+	}
+	return pvlc.gceCloudProvider, nil
+}
+
+func (pvlc *PersistentVolumeLabelController) updateVolume(volName string, volLabels map[string]string) error {
+	glog.V(4).Infof("updating PersistentVolume %s", volName)
+	curVol, err := pvlc.kubeClient.Core().PersistentVolumes().Get(volName, metav1.GetOptions{IncludeUninitialized: true})
```

Comment:
Alright, the logic has become more clear in here.  You're making the decision based on RV-A.  A conflicting patch will cause the controller to requeue and retry.  Is there actually a need for a live get here?  Why is that better than waiting for a watch event?  Is the previous decision valid on the new RV?